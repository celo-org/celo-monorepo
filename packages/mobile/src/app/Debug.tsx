import Button, { BtnTypes } from '@celo/react-components/components/Button'
import { anonymizedPhone } from '@celo/utils/src/phoneNumbers'
import * as React from 'react'
import { Clipboard, StyleSheet, Text, View } from 'react-native'
import DeviceInfo from 'react-native-device-info'
import { connect } from 'react-redux'
import LogView from 'src/app/LogView'
import { RootState } from 'src/redux/reducers'
import Logger from 'src/utils/Logger'
import { getLatestBlock } from 'src/web3/actions'
import { currentAccountSelector } from 'src/web3/selectors'

interface State {
  reactNativeLogs: string
  gethLogs: string
  latestBlockNumber: number
}

export class Debug extends React.Component<RootState, State> {
  static navigationOptions = { header: null }

  state = {
    reactNativeLogs: '',
    gethLogs: '',
    latestBlockNumber: 0,
  }

  async componentDidMount() {
    await this.updateLogs()
    await this.updateLatestBlock()
  }

  updateLogs = async () => {
    const logs = await Logger.getLogs()
    this.setState({
      reactNativeLogs: (logs && logs.reactNativeLogs) || 'Not Found',
      gethLogs: (logs && logs.gethLogs) || 'Not Found',
    })
  }

  updateLatestBlock = async () => {
    const latestBlock = await getLatestBlock()
    this.setState({
      latestBlockNumber: (latestBlock && latestBlock.number) || 0,
    })
  }

  onClickText = (...text: string[]) => {
    return () => {
      Logger.showMessage('Copied to Clipboard')
      Clipboard.setString(text.join(', '))
    }
  }

  onClickEmailLogs = async () => {
    await Logger.emailLogsToSupport(anonymizedPhone(this.props.account.e164PhoneNumber))
  }

  render() {
    const { reactNativeLogs, gethLogs, latestBlockNumber } = this.state
    const pincodeType = this.props.account.pincodeType
    const address = currentAccountSelector(this.props) || ''
    const phoneNumber = this.props.account.e164PhoneNumber
    const version = DeviceInfo.getVersion()
    const buildNumber = DeviceInfo.getBuildNumber()
    const apiLevel = DeviceInfo.getAPILevel()
    const deviceId = DeviceInfo.getDeviceId()
    return (
      <View style={style.container}>
        <Text
          onPress={this.onClickText(deviceId, phoneNumber)}
          style={style.singleLine}
        >{`Device Id: ${deviceId} | Phone Number: ${phoneNumber}`}</Text>
        <Text
          onPress={this.onClickText(version, buildNumber, String(apiLevel))}
          style={style.singleLine}
        >{`Version: ${version} | Build Number: ${buildNumber} | Api Level: ${apiLevel}`}</Text>
        <Text style={style.singleLine}>{`Pin Type: ${pincodeType}`}</Text>
        <Text
          onPress={this.onClickText(address)}
          style={style.singleLine}
        >{`Address: ${address}`}</Text>
        <Text style={style.singleLine}>{`Latest Block: ${latestBlockNumber}`}</Text>
        <LogView
          title={'React-Native Logs'}
          logs={reactNativeLogs}
          style={style.logView}
          onPress={this.onClickText(reactNativeLogs)}
        />
        <LogView
          title={'Geth Logs'}
          logs={gethLogs}
          style={style.logView}
          onPress={this.onClickText(gethLogs)}
        />
        <Button
          onPress={this.onClickEmailLogs}
          text={'Email logs to support'}
          standard={true}
          type={BtnTypes.PRIMARY}
          style={style.button}
        />
      </View>
    )
  }
}

const style = StyleSheet.create({
  container: {
    flex: 1,
    paddingVertical: 5,
    paddingHorizontal: 10,
    justifyContent: 'space-between',
    backgroundColor: '#FFFFFF',
  },
  singleLine: {
    marginTop: 5,
    fontSize: 11,
  },
  logView: {
    marginTop: 8,
    flex: 1,
  },
  button: {
    marginHorizontal: 0,
    marginTop: 10,
    marginBottom: 0,
  },
})

export default connect<RootState, {}, {}, RootState>((state: RootState) => state)(Debug)
