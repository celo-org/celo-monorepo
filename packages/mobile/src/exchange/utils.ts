import BigNumber from 'bignumber.js'
import { CURRENCY_ENUM } from 'src/geth/consts'
import { convertToContractDecimals } from 'src/tokens/saga'
import { getContractKitAsync } from 'src/web3/contracts'

export async function celoTocUsd(celoAmount: BigNumber) {
  const celoAmountInWei = (
    await convertToContractDecimals(celoAmount, CURRENCY_ENUM.GOLD)
  ).integerValue()

  const contractKit = await getContractKitAsync()
  const exchange = await contractKit.contracts.getExchange()
  const exchangeRate = await exchange.getGoldExchangeRate(celoAmountInWei)
  const cUsdAmount = exchangeRate.multipliedBy(celoAmount)

  return cUsdAmount
}
