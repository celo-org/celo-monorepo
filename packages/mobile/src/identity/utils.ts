import BigNumber from 'bignumber.js'

export function isUserBalanceSufficient(
  userBalance: string | number | null,
  estimatedTxAmount: number | string
) {
  if (!userBalance || new BigNumber(userBalance).isLessThan(new BigNumber(estimatedTxAmount))) {
    return false
  }

  return true
}
