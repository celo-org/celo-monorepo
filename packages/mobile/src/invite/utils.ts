import { sanitizeMessageBase64 } from '@celo/utils/src/attestations'
import URLSearchParamsReal from '@ungap/url-search-params'
import { Platform } from 'react-native'
import RNInstallReferrer from 'react-native-install-referrer'
import Logger from 'src/utils/Logger'

export const createInviteCode = (privateKey: string) => {
  // TODO(Rossy) we need some scheme to encrypt this PK
  // Buffer.from doesn't expect a 0x for hex input
  const privateKeyHex = privateKey.substring(2)
  return Buffer.from(privateKeyHex, 'hex').toString('base64')
}

// exported for testing
export const extractInviteCode = (inviteFieldInput: string) => {
  const sanitizedCode = sanitizeMessageBase64(inviteFieldInput)
  const regex = new RegExp('([0-9A-Za-z/\\+\\-\\_]*=)')
  const matches = sanitizedCode.match(regex)
  if (matches == null || matches.length === 0) {
    return null
  }
  return '0x' + Buffer.from(matches[0], 'base64').toString('hex')
}

// TODO(cmcewen): Consider web3 utils
export const isValidPrivateKey = (hexEncodedPrivateKey: string): boolean => {
  // First two chars are 0x
  if (hexEncodedPrivateKey.length !== 64 + 2) {
    return false
  }
  if (!hexEncodedPrivateKey.startsWith('0x')) {
    return false
  }
  return true
}

export function extractValidInviteCode(inviteFieldInput: string) {
  const inviteCode = extractInviteCode(inviteFieldInput)
  if (inviteCode == null || !isValidPrivateKey(inviteCode)) {
    return null
  } else {
    return inviteCode
  }
}

interface ReferrerData {
  clickTimestamp: string
  installReferrer: string
  installTimestamp: string
}

interface ReferrerDataError {
  message: string
}

export const getValidInviteCodeFromReferrerData = async () => {
  if (Platform.OS === 'android') {
    const referrerData: ReferrerData | ReferrerDataError = await RNInstallReferrer.getReferrer()
    Logger.info(
      'invite/utils/getInviteCodeFromReferrerData',
      'Referrer Data: ' + JSON.stringify(referrerData)
    )
    if (referrerData && referrerData.hasOwnProperty('installReferrer')) {
      const params = new URLSearchParamsReal(
        decodeURIComponent((referrerData as ReferrerData).installReferrer)
      )
      const inviteCode = params.get('invite-code')
      if (inviteCode) {
        const sanitizedCode = inviteCode.replace(' ', '+')
        // Accept invite codes which are either base64 encoded or direct hex keys
        if (isValidPrivateKey(sanitizedCode)) {
          return sanitizedCode
        }
        return extractValidInviteCode(sanitizedCode)
      }
    }
  }
  return null
}
