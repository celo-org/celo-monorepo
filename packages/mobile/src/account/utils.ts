import { ADDRESS_LENGTH } from 'src/exchange/reducer'

export const isAddressFormat = (content: string): boolean => {
  return content.startsWith('0x') && content.length === ADDRESS_LENGTH
}
