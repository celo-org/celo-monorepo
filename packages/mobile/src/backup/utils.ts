import { sampleSize } from 'lodash'
import { AsyncStorage } from 'react-native'
import { generateMnemonic, wordlists } from 'react-native-bip39'
import { getKey, setKey } from 'src/utils/keyStore'
import Logger from 'src/utils/Logger'

const TAG = 'Backup/utils'

export async function createQuizWordList(mnemonic: string, language: string | null) {
  const disallowedWordSet = new Set(mnemonic.split(' '))
  const languageWordList = getWordlist(language)
  const wordOptions: string = await generateMnemonic(1000, null, languageWordList)
  const quizWordList = new Set(
    [...wordOptions.split(' ')].filter((word: string) => !disallowedWordSet.has(word))
  )
  return [...quizWordList]
}

export function selectQuizWordOptions(correctWord: string, allWords: string[], numOptions: number) {
  const wordOptions = []
  const correctWordPosition = Math.floor(Math.random() * numOptions)
  const randomWordIndexList = sampleSize([...Array(allWords.length).keys()], numOptions - 1)
  let randomWordIndex: number = 0

  for (let i = 0; i < numOptions; i++) {
    if (i === correctWordPosition) {
      wordOptions.push(correctWord)
      continue
    }

    wordOptions.push(allWords[randomWordIndexList[randomWordIndex]])
    randomWordIndex += 1
  }
  return wordOptions
}

export function getWordlist(language: string | null) {
  if (!language) {
    return wordlists.EN
  }

  switch (language.slice(0, 2)) {
    case 'es': {
      return wordlists.ES
    }
    default: {
      return wordlists.EN
    }
  }
}

// TODO(Rossy) Remove after the next alfa testnet reset
export async function getStoredMnemonic(): Promise<string | null> {
  try {
    Logger.debug(TAG, 'Checking keystore for mnemonic')
    let mnemonic = await getKey('mnemonic')
    if (mnemonic) {
      return mnemonic
    }

    Logger.debug(TAG, 'Mnemonic not found in keystore, checking async storage')
    mnemonic = await AsyncStorage.getItem('mnemonic')
    if (mnemonic) {
      await setKey('mnemonic', mnemonic)
      await AsyncStorage.removeItem('mnemonic')
      return mnemonic
    }

    Logger.error(TAG, 'No mnemonic found')
    return null
  } catch (error) {
    Logger.error(TAG, 'Failed to retrieve mnemonic', error)
    return null
  }
}
