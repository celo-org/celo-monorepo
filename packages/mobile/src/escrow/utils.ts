import { ensureLeading0x } from '@celo/base'
import { generateDeterministicInviteCode } from '@celo/utils/lib/account'
import { publicKeyToAddress } from '@celo/utils/lib/address'

export const generateEscrowPaymentIdAndPk = (
  recipientPhoneHash: string,
  recipientPepper: string,
  addressIndex: number = 0
) => {
  const { publicKey, privateKey } = generateDeterministicInviteCode(
    recipientPhoneHash,
    recipientPepper,
    addressIndex
  )
  return { paymentId: publicKeyToAddress(publicKey), privateKey: ensureLeading0x(privateKey) }
}

export const generateUniquePaymentId = (
  existingPaymentIds: string[],
  phoneHash: string,
  pepper: string
) => {
  const paymentIdSet: Set<string> = new Set(existingPaymentIds)

  // Using an upper bound of 100 to be sure this doesn't run forever given
  // the realistic amount of pending escrow txs is far less than this
  for (let i = 0; i < 100; i += 1) {
    const { paymentId } = generateEscrowPaymentIdAndPk(phoneHash, pepper, i)
    if (!paymentIdSet.has(paymentId)) {
      return paymentId
    }
  }
}
