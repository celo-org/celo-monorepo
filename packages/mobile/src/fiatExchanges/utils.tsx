import { SIMPLEX_URI } from 'src/config'
import { CURRENCY_ENUM } from 'src/geth/consts'
import { LocalCurrencyCode } from 'src/localCurrency/consts'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { navigateToURI } from 'src/utils/linking'

export const openMoonpay = (currencyCode: LocalCurrencyCode, currencyToBuy: CURRENCY_ENUM) => {
  navigate(Screens.MoonPay, {
    localAmount: 0,
    currencyCode,
    currencyToBuy,
  })
}

export const openSimplex = (account: string | null) => {
  navigateToURI(`${SIMPLEX_URI}?address=${account}`)
}
