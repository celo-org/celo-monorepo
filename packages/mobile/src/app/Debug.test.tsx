import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import * as renderer from 'react-test-renderer'
import Debug from 'src/app/Debug'
import { createMockStore } from 'test/utils'

jest.mock('src/web3/contracts', () => ({
  isZeroSyncMode: jest.fn().mockReturnValueOnce(false),
}))

describe('Debug', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <Provider store={createMockStore()}>
        <Debug />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })
})
