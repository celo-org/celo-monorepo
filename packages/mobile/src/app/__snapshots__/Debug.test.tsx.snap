// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Debug renders correctly 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "space-between",
      "paddingHorizontal": 10,
      "paddingVertical": 5,
    }
  }
>
  <Text
    onPress={[Function]}
    style={
      Object {
        "fontSize": 11,
        "marginTop": 5,
      }
    }
  >
    Device Id: undefined | Phone Number: +14155556666
  </Text>
  <Text
    onPress={[Function]}
    style={
      Object {
        "fontSize": 11,
        "marginTop": 5,
      }
    }
  >
    Version: undefined | Build Number: undefined | Api Level: undefined
  </Text>
  <Text
    style={
      Object {
        "fontSize": 11,
        "marginTop": 5,
      }
    }
  >
    Pin Type: Unset
  </Text>
  <Text
    onPress={[Function]}
    style={
      Object {
        "fontSize": 11,
        "marginTop": 5,
      }
    }
  >
    Address: 0x0000000000000000000000000000000000007E57
  </Text>
  <Text
    style={
      Object {
        "fontSize": 11,
        "marginTop": 5,
      }
    }
  >
    Latest Block: 0
  </Text>
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
        Object {
          "flex": 1,
          "marginTop": 8,
        },
      ]
    }
  >
    <Text>
      React-Native Logs
    </Text>
    <RCTScrollView
      onContentSizeChange={[Function]}
    >
      <View>
        <Text
          onPress={[Function]}
          style={
            Object {
              "fontSize": 10,
            }
          }
        >
          
        </Text>
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
        Object {
          "flex": 1,
          "marginTop": 8,
        },
      ]
    }
  >
    <Text>
      Geth Logs
    </Text>
    <RCTScrollView
      onContentSizeChange={[Function]}
    >
      <View>
        <Text
          onPress={[Function]}
          style={
            Object {
              "fontSize": 10,
            }
          }
        >
          
        </Text>
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 3,
          "flexDirection": "row",
        },
        Object {
          "marginVertical": 10,
        },
        Object {
          "marginBottom": 10,
        },
        Object {
          "marginBottom": 0,
          "marginHorizontal": 0,
          "marginTop": 10,
        },
        Object {
          "backgroundColor": "#42D689",
        },
      ]
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      nativeBackgroundAndroid={
        Object {
          "attribute": "selectableItemBackground",
          "type": "ThemeAttrAndroid",
        }
      }
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          },
          Object {
            "backgroundColor": "#42D689",
          },
          Object {
            "height": 50,
          },
          Object {
            "borderColor": "#42D689",
            "borderRadius": 3,
            "borderWidth": 2,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Hind-SemiBold",
                "fontSize": 16,
              },
              Object {
                "color": "#FFFFFF",
              },
              Object {
                "paddingLeft": 5,
                "paddingRight": 5,
              },
            ]
          }
        >
          Email logs to support
        </Text>
      </View>
    </View>
  </View>
</SafeAreaView>
`;
